server/src/server.ts

```diff
@@ -670,6 +707,183 @@ connection.onSignatureHelp((data) => {
+      const { struct, id } = searchResult.data;
+      if (struct.loc?.source?.endsWith('.h0') || struct.loc?.source?.endsWith('.h1')) {
+        return new ResponseError<void>(0, "Can't rename that field: It's defined in a library.");
+      }
+      if (id.loc) return ast.locToRange(id.loc);
+      break;
+    }
+  }
+
+  // Generic invalid rename
+  return new ResponseError<void>(0, "Can't rename that.");
+});
+
+connection.onRenameRequest(async (data) => {
+  const genv = openFiles.get(data.textDocument.uri);
+  // Indicates no successful parse so far
+  if (genv === undefined) { return null; }
+
+  const pos: ast.Position = ast.fromVscodePosition(data.position);
+  const searchResult = findGenv({ pos, genv }, data.textDocument.uri);
+
+  // This indicates that the user hovered over something that
+  // wasn't an indentifier
+  if (searchResult === null || searchResult.data === null) return null;
+
+  let toFind: FindUsesParam | null = null;
+
+  switch (searchResult.data.tag) {
+    case "FoundType": {
+      const { type } = searchResult.data;
+
+      switch (type.tag) {
+        case "Identifier":
+          toFind = {
+            tag: 'FindType',
+            name: type.name
+          };
+          break;
+
+        case "StructType":
+          toFind = {
+            tag: 'FindStruct',
+            name: type.id.name
+          };
+          break;
+      }
+      break;
+    }
+    case "FoundIdent": {
+      const { name, type } = searchResult.data;
+
+      if (type.tag === "FunctionType") {
+        toFind = {
+          tag: 'FindFunction',
+          name: name
+        };
+        break;
+      } else {
+        const definition: EnvEntry | undefined = searchResult.environment?.get(name);
+  
+        if (definition !== undefined && definition.position !== undefined) {
+          toFind = {
+            tag: 'FindVar',
+            entry: definition
+          };
+        }
+      }
+      break;
+    }
+    case "FoundField": {
+      const { struct, field } = searchResult.data;
+      
+      toFind = {
+        tag: 'FindField',
+        struct: struct.id.name,
+        field: field.id.name
+      };
+      break;
+    }
+  }
+  if (toFind) {
+    // Get all files which could use this file
+    const dir = path.dirname(data.textDocument.uri);
+    const folders = await connection.workspace.getWorkspaceFolders();
+    const files = getAllFiles(data.textDocument.uri, getConfigPaths(dir, folders));
+    try {
+      const changes: {
+        [uri: string]: TextEdit[]
+      } = { };
+      for (const file of files) {
+        let fileGenv = openFiles.get(file);
+        if (!fileGenv) {
+          validateTextDocument(file);
+          fileGenv = openFiles.get(file);
+          if (!fileGenv) {
+            // Give up with this file
+            continue;
+          }
+        }
+        changes[file] = findUses(fileGenv, file, toFind).map((loc) => {
+          return TextEdit.replace(ast.locToRange(loc), data.newName);
```

Comment:
Should we try to validate that `data.newName` is a valid identifier? Just need to test it against `basicLexing.identifier.match`

Reply:
Good point. Will do